code/modules/assembly/mousetrap.dm

```diff
@@ -46,7 +46,7 @@
 			return FALSE
 		switch(type)
 			if("feet")
-				if(!H.shoes || !(H.shoes.body_parts_covered & FEET))
+				if(!HAS_TRAIT(H,TRAIT_HARDSOLES) && (!H.shoes || !(H.shoes.body_parts_covered & FEET)))
```

Comment:
move the trait check two lines below, as well as add another walk move_intent check on that line.